docs(spec): add breadcrumb spec

khamudom opened the pull request:
<!--- Provide a summary of your changes in the title field above. For guidance on formatting, see the comment at the bottom of this template. -->

# Description

The file has specs for `fast-breadcrumb` and `fast-breadcrumb-item`

![image](https://user-images.githubusercontent.com/37851220/88443820-ae553c00-cdce-11ea-9ab9-14d1482eab3e.png)

## Motivation & context

<!--- What problem does this change solve? -->
<!--- Provide a link if you are addressing an open issue. -->

## Issue type checklist

<!--- What type of change are you submitting? Put an x in the box that applies: -->

- [ ] **Chore**: A change that does not impact distributed packages.
- [ ] **Bug fix**: A change that fixes an issue, link to the issue above.
- [ ] **New feature**: A change that adds functionality.

**Is this a breaking change?**
- [ ] This change causes current functionality to break.

<!--- If yes, describe the impact. -->

**Adding or modifying component(s) in `@microsoft/fast-components` checklist**

<!-- Do your changes add or modify components in the @microsoft/fast-components package? Put an x in the box that applies: -->

- [ ] I have added a new component
- [ ] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [ ] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

## Process & policy checklist

<!--- Review the list and check the boxes that apply. -->

- [ ] I have added tests for my changes.
- [ ] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/en/contributing/standards) for this project.

<!---
Formatting guidelines:

Accepted peer review title format:
<type>: <description>

Example titles:
    chore: add unit tests for all components
    feat: add a border radius to button
    fix: update design system to use 3px border radius

    <type> is required to be one of the following:

        - chore: A change that does not impact distributed packages.
        - fix: A change which fixes an issue.
        - feat: A that adds functionality.

    <description> is required for the CHANGELOG and speaks to what the user gets from this PR:

        - Be concise.
        - Use all lowercase characters. 
        - Use imperative, present tense (e.g. `add` not `adds`.)
        - Do not end your description with a period.
        - Avoid redundant words.

For additional information regarding working on FAST, check out our documentation site:
https://www.fast.design/docs/en/contributing/working
-->

chrisdholt commented on specs/breadcrumb.md:
````diff
@@ -0,0 +1,132 @@
+
+```html
+<fast-breadcrumb>
+    <fast-breadcrumb-item>
+        <fast-anchor appearance="lightweight">Home</fast-anchor>
+    </fast-breadcrumb-item>
+    <fast-breadcrumb-item>
+        <fast-anchor appearance="lightweight">Templates</fast-anchor>
+    </fast-breadcrumb-item>
+    <fast-breadcrumb-item aria-current="page">
+        <fast-anchor appearance="lightweight">Template Prices</fast-anchor>
+    </fast-breadcrumb-item>
````
Is there a reason to have `fast-breadcrumb-item` if it's going to include fast anchors?

khamudom replied:
`fast-anchor` is just a component you can use as a breadcrumb item. breadcrumb-item has a default slot to implement any control and also has the separator.

chrisdholt replied:
In the current spec that doesn't seem to be the case; I need to slot fast-anchor into the fast-breadcrumb-item, which seems like an unnecessary abstraction here.

khamudom replied:
I agree, does feel unnecessary. The issue I am currently running into is a way to inject the separator between each item and also have separator be any element. That is why we have breadcrumb-item. 
I am going to keep investigating this so we don't need a breadcrumb-item.

chrisdholt commented on specs/breadcrumb.md:
```diff
@@ -0,0 +1,132 @@
+
+*Slots*
+- default slot for element
+- `separator` - used as part of the visual presentation that signifies the breadcrumb trail. The separator is defaulted to a `/`, but authors can override to use an `svg`.
```
Should this be in the item or in `fast-breadcrumb`?

chrisdholt commented on specs/breadcrumb.md:
````diff
@@ -0,0 +1,132 @@
+### Anatomy and Appearance
+
+```ts
+<div
+    role="listitem"
+    class="content"
+    part="content"
+    tabindex="0"
+    aria-current="${x => x.ariaCurrent}"
+>
````
I'm not sure we want to focus on the list item if the focusable element is within this. Additionally I think we'd want aria-current on the `link` that's slotted in here. This seems extraneous and potentially unnecessary if the primary goal is to serve as a wrapper for the link and a visual separator.

Is there a way to simplify here? Do we need this layer?

khamudom replied:
updated to set aria-current on the control inside the breadcrumb item.

chrisdholt replied:
We should call out that aria-current is *only* necessary if the element is a link. If this doesn't have an href it'll be read incorrectly by AT.

khamudom replied:
I'll update the accessibility section explaining this. Thanks

EisenbergEffect commented:
@khamudom and I worked through a lot of options and there was definitely a difficulty in dealing with getting to right structure and incorporating the separator. It seems like the only way to get the separator is to have a wrapper element to render it because there's no way to wrap individual slotted elements.

As I was reading the comments I did come up with a new idea though. Caveat: it's dependent on the new work in fast-element and the proposed new FoundationElement base class. With the new base class, all of our components would have a `template` property to allow for dynamically swapping the template of an element. With that in place, we could do something really interesting...we could have the breadcrumb component grab the template of each of its children, and then create a new template that wraps the current template so that it can apply the configured separator. It can then assign the new composed template back to the anchor instances. What do you all think of that idea?

chrisdholt commented:
@EisenbergEffect I was rolling around something like that in my head but wasn’t sure how to reconcile with composability. I think that would be preferred here and a welcome addition. It’s worth waiting for that work in my opinion.

EisenbergEffect commented:
There are a few more tricks to it since the child anchors would not be connected when the breadcrumbs connected callback fires. So, the child templates wouldn't be in place yet to wrap. I've got a couple of solutions in mind to it and will keep thinking about it a bit more. It might be worth introducing "template decorators" as an official concept for things like this and have some standard way of doing it. Need to think about it a bit more.

radium-v commented:
Would [css counter styles](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Lists_and_Counters/Using_CSS_counters) work for the separators?

EisenbergEffect commented:
@radium-v I'm not sure. It probably depends on how flexible we want to make the separator. What if someone has a custom SVG, for example, do we want to support that or just a basic character?

radium-v commented:
@EisenbergEffect my bad, I was thinking about [`list-style` properties](https://developer.mozilla.org/en-US/docs/Web/CSS/list-style) which do allow for images. Interestingly, the [breadcrumb example linked in the README](https://www.w3.org/TR/wai-aria-practices/examples/breadcrumb/index.html) uses empty `::before` pseudo-elements to avoid excessive verbosity with screen readers.

EisenbergEffect commented:
Is it possible to apply a list style to something that isn't actually an li? If we give it an aria `listitem` role, would that styling work? I'd certainly prefer that type of solution if possible.

nicholasrice commented on specs/breadcrumb.md:
```diff
@@ -0,0 +1,135 @@
+
+ *Attributes:*
+- `aria-current` - aria attribute set on the last item of breadcrumb.
+- `separator` - a divider between two breadcrumb items. The default of this attribute is `/`.
```
Thoughts on allowing this to be a `html` template instance? That would provide authors far more options than just a content character

EisenbergEffect replied:
In this case do you mean a vanilla template (not a FAST template with bindings)?

khamudom replied:
removed the `separator` attribute. This was to set a default separator. Will used the slot instead to set a default.

janechu commented on specs/breadcrumb.md:
```diff
@@ -0,0 +1,135 @@
+- `fast-breadcrumb-item`
+
+ *Attributes:*
+- `aria-current` - aria attribute set on the last item of breadcrumb.
```
Should it be noted here that this is always `aria-current="page"`?

khamudom replied:
Will not that.

janechu commented on specs/breadcrumb.md:
```diff
@@ -0,0 +1,135 @@
+>
+    <slot></slot>
+    <span
+        class="separator ${x => (x.ariaCurrent ? "hide" : "")}
```
```suggestion
        class="separator ${x => (x.ariaCurrent ? "hide" : "")}"
```

khamudom replied:
Thanks

khamudom commented:
Updated the fast-breadcrumb-item. instead of having an empty slot, added am anchor as a default control. This cleans up the html, it avoids having to add another control inside fast-breadcrumb-item.

nicholasrice commented on specs/breadcrumb.md:
````diff
@@ -0,0 +1,139 @@
+```html
+<fast-breadcrumb>
+    <fast-breadcrumb-item name="Home" href="#"></fast-breadcrumb-item>
+    <fast-breadcrumb-item name="Template" href="#"></fast-breadcrumb-item>
````
I like this a lot better - not needing to add an an anchor inside. nice!

chrisdholt commented on specs/breadcrumb.md:
```diff
@@ -0,0 +1,139 @@
+
+### Globalization
+
+The first and last element inside the breadcrumb will be reversed when switching in LTR/RTL.
```
Nit: I think this would be the overall order is reversed. If there are only three items, this is true - if there are six, not as much :).

khamudom replied:
Good point. Thanks

chrisdholt commented on specs/breadcrumb.md:
````diff
@@ -0,0 +1,139 @@
+```html
+<div
+    role="listitem"
+    class="listitem-container"
+    part="listitem-container"
````
Just an FYI to catch this prior to implementation...I think this can just be `listitem`; it doesn't contain the list item, it's serving as the list item.

chrisdholt commented on specs/breadcrumb.md:
```diff
@@ -0,0 +1,139 @@
+    ${when(
+        x => x.showSeparator,
+        html<BreadcrumbItem>`
+            <span class="separator" aria-hidden="true">
```
You'll want the part here where you have the class.

chrisdholt commented on specs/breadcrumb.md:
```diff
@@ -0,0 +1,139 @@
+- `fast-breadcrumb-item`
+
+*Properties*
+- `showSeparator` - is a boolean to show and hide the separator.
```
If I'm using fast-breadcrumb with these items, do I need to set this on every item if I want to change from the default (assuming the default is true, I want to hide it and set false)? Or is the intent with this being a property that we have a method of hiding all nested fast-breadcrumb-items separators if we want?

khamudom replied:
Yes, `showSeparator` is true by default. We can create a method that will hide all separator, currently, my thought was to only hide the separator on the last item. In Breadcrumb, I will use the `instanceof` BreadcrumbItem, and on the last item, set `showSeparator` to false. 
You also brought up a point on changing separator on each item from default. I will have to come up with a way to only change the separator once and not each item.

khamudom replied:
Yes, `showSeparator` is defaulted true. In Breadcrumb, I would use `instanceof` BreadcrumbItem and set `showSeparator` on the last item to false. Since the separator visually represents a breadcrumb trail, I did not think to hide all of them.
To change the default of the separator, I will have to come up with a way to change it once, and not have to change it on all the items.

chrisdholt commented on specs/breadcrumb.md:
````diff
@@ -0,0 +1,139 @@
+```html
+<template
+    role="navigation"
+    aria-label="breadcrumb"
````
Is this aria-label part of the implementation perhaps? I'm assuming you don't intend to hardcode this as it needs to localize. Additionally, I believe that someone could also use aria-labelledby in place of the aria-label. How do we anticipate supporting both? Should it be guidance or should we have first class attributes for these? One _is_ required: https://www.w3.org/TR/wai-aria-practices-1.2/#wai-aria-roles-states-and-properties-2

khamudom replied:
I will create attributes for both `aria-label` and `aria-labelledby`
